File: tools/hod2lib/texbank.py
"""
HOTD2 texture banks.

See docs/formats/texbank.md.

A tex/ file is the raw PowerVR2 texture payloads concatenated in texture-ID
order, with no header, no padding and no alignment. All metadata -- width,
height, pixel format, VQ, twiddling -- lives in the *model* that references the
texture, in its TSP and texture_control words.

Empirically, across the whole game:
  * pixel formats are only RGB565, ARGB4444 and ARGB1555 (all 16-bit)
  * there are no palettised textures, so no palette data exists anywhere
  * there are no mipmaps
  * VQ and plain twiddled/linear layouts both occur
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

__all__ = ["TexDesc", "decode", "harvest_descriptors", "solve_layout", "Bank"]

VQ_CODEBOOK_BYTES = 2048  # 256 entries x 2x2 pixels x 2 bytes


@dataclass(frozen=True)
class TexDesc:
    """Everything needed to locate and decode one texture."""

    width: int
    height: int
    pixfmt: int
    vq: bool
    mipmap: bool
    twiddled: bool

    @classmethod
    def from_mesh(cls, mesh) -> "TexDesc":
        return cls(
            width=mesh.texture_width,
            height=mesh.texture_height,
            pixfmt=mesh.pixel_format,
            vq=mesh.vq_compressed,
            mipmap=mesh.mipmapped,
            twiddled=mesh.twiddled,
        )

    @property
    def size(self) -> int:
        """Bytes this texture occupies in the bank."""
        if self.vq:
            return VQ_CODEBOOK_BYTES + (self.width * self.height) // 4
        bpp = 4 if self.pixfmt == 5 else 8 if self.pixfmt == 6 else 16
        return (self.width * self.height * bpp) // 8


# --------------------------------------------------------------------------
# Pixel formats. All 16-bit little-endian; output is RGBA8888.
# --------------------------------------------------------------------------

def _argb1555(p: int) -> tuple[int, int, int, int]:
    a = 255 if (p >> 15) & 1 else 0
    r = ((p >> 10) & 0x1F) * 255 // 31
    g = ((p >> 5) & 0x1F) * 255 // 31
    b = (p & 0x1F) * 255 // 31
    return r, g, b, a


def _rgb565(p: int) -> tuple[int, int, int, int]:
    r = ((p >> 11) & 0x1F) * 255 // 31
    g = ((p >> 5) & 0x3F) * 255 // 63
    b = (p & 0x1F) * 255 // 31
    return r, g, b, 255


def _argb4444(p: int) -> tuple[int, int, int, int]:
    a = ((p >> 12) & 0xF) * 17
    r = ((p >> 8) & 0xF) * 17
    g = ((p >> 4) & 0xF) * 17
    b = (p & 0xF) * 17
    return r, g, b, a


_DECODERS = {0: _argb1555, 1: _rgb565, 2: _argb4444, 5: _argb1555, 6: _argb1555}


# --------------------------------------------------------------------------
# Twiddling (Morton order)
# --------------------------------------------------------------------------

def _morton(x: int, y: int) -> int:
    """Interleave bits: y at even positions, x at odd."""
    v = 0
    for i in range(16):
        v |= ((y >> i) & 1) << (2 * i)
        v |= ((x >> i) & 1) << (2 * i + 1)
    return v


def twiddled_index(x: int, y: int, w: int, h: int) -> int:
    """Index into twiddled data for pixel (x, y).

    Non-square textures are handled as a run of square blocks of side
    min(w, h), laid out linearly along the longer axis.
    """
    if w == h:
        return _morton(x, y)
    if w > h:
        return (x // h) * h * h + _morton(x % h, y)
    return (y // w) * w * w + _morton(x, y % w)


# --------------------------------------------------------------------------
# Decode
# --------------------------------------------------------------------------

def decode(data: bytes, off: int, d: TexDesc) -> bytearray:
    """Decode one texture to RGBA8888 (width*height*4 bytes)."""
    w, h = d.width, d.height
    conv = _DECODERS.get(d.pixfmt, _rgb565)
    out = bytearray(w * h * 4)

    if d.vq:
        # 2048-byte codebook of 256 entries, each a 2x2 pixel block.
        cb_end = off + VQ_CODEBOOK_BYTES
        book = struct.unpack_from("<1024H", data, off)
        idx = data[cb_end : cb_end + (w * h) // 4]

        bw, bh = w // 2, h // 2
        for by in range(bh):
            for bx in range(bw):
                # Index data is itself twiddled, at half resolution.
                i = idx[twiddled_index(bx, by, bw, bh)]
                blk = book[i * 4 : i * 4 + 4]
                # Column-major within the 2x2 block.
                for k, (dx, dy) in enumerate(((0, 0), (0, 1), (1, 0), (1, 1))):
                    px, py = bx * 2 + dx, by * 2 + dy
                    r, g, b, a = conv(blk[k])
                    o = (py * w + px) * 4
                    out[o : o + 4] = bytes((r, g, b, a))
        return out

    px16 = struct.unpack_from("<%dH" % (w * h), data, off)
    for y in range(h):
        for x in range(w):
            src = twiddled_index(x, y, w, h) if d.twiddled else y * w + x
            r, g, b, a = conv(px16[src])
            o = (y * w + x) * 4
            out[o : o + 4] = bytes((r, g, b, a))
    return out


# --------------------------------------------------------------------------
# Bank layout
# --------------------------------------------------------------------------

def harvest_descriptors(models) -> dict[int, TexDesc]:
    """Collect texture_id -> TexDesc from parsed models."""
    out: dict[int, TexDesc] = {}
    for m in models:
        for mesh in m.meshes:
            if mesh.texture_id >= 0:
                out.setdefault(mesh.texture_id, TexDesc.from_mesh(mesh))
    return out


@dataclass
class Bank:
    data: bytes
    descs: dict[int, TexDesc]
    offsets: dict[int, int]
    complete: bool
    residual: int

    def decode(self, tex_id: int) -> tuple[int, int, bytearray] | None:
        d = self.descs.get(tex_id)
        off = self.offsets.get(tex_id)
        if d is None or off is None:
            return None
        if off + d.size > len(self.data):
            return None
        return d.width, d.height, decode(self.data, off, d)


def bank_from_exe(data: bytes, entries) -> Bank:
    """Build a Bank from the authoritative descriptor table in Hod2.exe.

    Always prefer this over solve_layout: the exe carries exact offsets,
    dimensions and layout codes, whereas a prefix sum cannot reproduce the
    2048-byte padding or the VQ half-size convention.
    """
    descs: dict[int, TexDesc] = {}
    offsets: dict[int, int] = {}
    for e in entries:
        descs[e.index] = TexDesc(
            width=e.width, height=e.height, pixfmt=e.pixfmt,
            vq=e.vq, mipmap=False, twiddled=e.twiddled,
        )
        offsets[e.index] = e.offset

    used = 0
    if entries:
        last = max(entries, key=lambda e: e.offset)
        used = last.offset + descs[last.index].size
    return Bank(data, descs, offsets, used <= len(data), len(data) - used)


def solve_layout(data: bytes, descs: dict[int, TexDesc]) -> Bank:
    """Compute the byte offset of each texture in a bank.

    Textures are concatenated in ID order, so offsets are a prefix sum -- but
    only if every ID from 0..max is known. A gap (an ID present in the bank but
    referenced by no model we parsed) makes every later offset unknown.

    Returns a Bank with whatever could be resolved. ``complete`` is True when
    the prefix sum covers every ID and lands exactly on the file size.
    """
    if not descs:
        return Bank(data, descs, {}, False, len(data))

    top = max(descs)
    offsets: dict[int, int] = {}
    pos = 0
    ok = True
    for i in range(top + 1):
        d = descs.get(i)
        if d is None:
            ok = False
            break
        offsets[i] = pos
        pos += d.size

    residual = len(data) - pos
    complete = ok and residual == 0
    return Bank(data, descs, offsets, complete, residual)

File: web/src/hod2lib/texbank.ts
/**
 * HOTD2 texture banks. The port of `tools/hod2lib/texbank.py`.
 *
 * See docs/formats/texbank.md.
 *
 * A tex/ file is the raw PowerVR2 texture payloads concatenated in texture-ID
 * order, with no header, no padding and no alignment. All metadata -- width,
 * height, pixel format, VQ, twiddling -- lives in the *model* that references
 * the texture, in its TSP and texture_control words.
 *
 * Empirically, across the whole game:
 *   * pixel formats are only RGB565, ARGB4444 and ARGB1555 (all 16-bit)
 *   * there are no palettised textures, so no palette data exists anywhere
 *   * there are no mipmaps
 *   * VQ and plain twiddled/linear layouts both occur
 */

import { u16s } from "./bytes";
import type { Mesh, Model } from "./nl1";

/** 256 entries x 2x2 pixels x 2 bytes. */
export const VQ_CODEBOOK_BYTES = 2048;

/** Everything needed to locate and decode one texture. */
export interface TexDesc {
  width: number;
  height: number;
  pixfmt: number;
  vq: boolean;
  mipmap: boolean;
  twiddled: boolean;
}

export function descFromMesh(mesh: Mesh): TexDesc {
  return {
    width: mesh.textureWidth,
    height: mesh.textureHeight,
    pixfmt: mesh.pixelFormat,
    vq: mesh.vqCompressed,
    mipmap: mesh.mipmapped,
    twiddled: mesh.twiddled,
  };
}

/** Bytes this texture occupies in the bank. */
export function descSize(d: TexDesc): number {
  if (d.vq) return VQ_CODEBOOK_BYTES + Math.floor((d.width * d.height) / 4);
  const bpp = d.pixfmt === 5 ? 4 : d.pixfmt === 6 ? 8 : 16;
  return Math.floor((d.width * d.height * bpp) / 8);
}

// ---------------------------------------------------------------------------
// Pixel formats. All 16-bit little-endian; output is RGBA8888.
// ---------------------------------------------------------------------------

/**
 * Each decoder writes four bytes into *out* rather than returning a tuple.
 *
 * The reference implementation returns `(r, g, b, a)` and the caller unpacks
 * it. That is one tuple allocation per pixel, and a stage decodes about
 * seventeen million of them.
 */
type Decoder = (p: number, out: Uint8Array, o: number) => void;

const argb1555: Decoder = (p, out, o) => {
  out[o] = (((p >> 10) & 0x1f) * 255 / 31) | 0;
  out[o + 1] = (((p >> 5) & 0x1f) * 255 / 31) | 0;
  out[o + 2] = ((p & 0x1f) * 255 / 31) | 0;
  out[o + 3] = (p >> 15) & 1 ? 255 : 0;
};

const rgb565: Decoder = (p, out, o) => {
  out[o] = (((p >> 11) & 0x1f) * 255 / 31) | 0;
  out[o + 1] = (((p >> 5) & 0x3f) * 255 / 63) | 0;
  out[o + 2] = ((p & 0x1f) * 255 / 31) | 0;
  out[o + 3] = 255;
};

const argb4444: Decoder = (p, out, o) => {
  out[o] = ((p >> 8) & 0xf) * 17;
  out[o + 1] = ((p >> 4) & 0xf) * 17;
  out[o + 2] = (p & 0xf) * 17;
  out[o + 3] = ((p >> 12) & 0xf) * 17;
};

const DECODERS: Record<number, Decoder> = {
  0: argb1555, 1: rgb565, 2: argb4444, 5: argb1555, 6: argb1555,
};

// ---------------------------------------------------------------------------
// Twiddling (Morton order)
// ---------------------------------------------------------------------------

/**
 * `spread[v]` is *v* with its bits moved to the even positions.
 *
 * The reference implementation interleaves in a sixteen-iteration loop per
 * pixel. A texture is up to 1024x1024 and a stage has 1,730 of them, so the
 * loop is the decode; a table of 1024 entries is the same arithmetic done
 * once. The largest texture dimension in the game is 1024, and
 * {@link morton} asserts nothing beyond that because a wider one would index
 * past the end of a bank long before it produced a wrong pixel.
 */
const SPREAD = (() => {
  const t = new Uint32Array(1024);
  for (let v = 0; v < 1024; v++) {
    let s = 0;
    for (let i = 0; i < 10; i++) s |= ((v >> i) & 1) << (2 * i);
    t[v] = s;
  }
  return t;
})();

/** Interleave bits: y at even positions, x at odd. */
export function morton(x: number, y: number): number {
  return SPREAD[y] | (SPREAD[x] << 1);
}

/**
 * Index into twiddled data for pixel (x, y).
 *
 * Non-square textures are handled as a run of square blocks of side
 * min(w, h), laid out linearly along the longer axis.
 */
export function twiddledIndex(x: number, y: number, w: number,
                              h: number): number {
  if (w === h) return morton(x, y);
  if (w > h) return Math.floor(x / h) * h * h + morton(x % h, y);
  return Math.floor(y / w) * w * w + morton(x, y % w);
}

/** Decode one texture to RGBA8888 (`width * height * 4` bytes). */
export function decode(data: Uint8Array, off: number,
                       d: TexDesc): Uint8Array {
  const w = d.width, h = d.height;
  const conv = DECODERS[d.pixfmt] ?? rgb565;
  const out = new Uint8Array(w * h * 4);

  if (d.vq) {
    // 2048-byte codebook of 256 entries, each a 2x2 pixel block.
    const cbEnd = off + VQ_CODEBOOK_BYTES;
    const book = u16s(data, off, 1024);
    const idx = data.subarray(cbEnd, cbEnd + Math.floor((w * h) / 4));

    const bw = w >> 1, bh = h >> 1;
    // Column-major within the 2x2 block.
    const DX = [0, 0, 1, 1];
    const DY = [0, 1, 0, 1];
    for (let by = 0; by < bh; by++) {
      for (let bx = 0; bx < bw; bx++) {
        // Index data is itself twiddled, at half resolution.
        const i = idx[twiddledIndex(bx, by, bw, bh)];
        for (let k = 0; k < 4; k++) {
          const px = bx * 2 + DX[k];
          const py = by * 2 + DY[k];
          conv(book[i * 4 + k], out, (py * w + px) * 4);
        }
      }
    }
    return out;
  }

  const px16 = u16s(data, off, w * h);
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const src = d.twiddled ? twiddledIndex(x, y, w, h) : y * w + x;
      conv(px16[src], out, (y * w + x) * 4);
    }
  }
  return out;
}

/** Collect `textureId -> TexDesc` from parsed models. */
export function harvestDescriptors(models: Model[]): Map<number, TexDesc> {
  const out = new Map<number, TexDesc>();
  for (const m of models) {
    for (const mesh of m.meshes) {
      if (mesh.textureId >= 0 && !out.has(mesh.textureId)) {
        out.set(mesh.textureId, descFromMesh(mesh));
      }
    }
  }
  return out;
}

export interface Bank {
  data: Uint8Array;
  descs: Map<number, TexDesc>;
  offsets: Map<number, number>;
  complete: boolean;
  residual: number;
}

export function bankDecode(bank: Bank, texId: number):
    { width: number; height: number; pixels: Uint8Array } | null {
  const d = bank.descs.get(texId);
  const off = bank.offsets.get(texId);
  if (d === undefined || off === undefined) return null;
  if (off + descSize(d) > bank.data.length) return null;
  return { width: d.width, height: d.height,
           pixels: decode(bank.data, off, d) };
}

/** One entry of the exe's texture descriptor table. See `exetab`. */
export interface TexEntryLike {
  index: number;
  offset: number;
  width: number;
  height: number;
  pixfmt: number;
  vq: boolean;
  twiddled: boolean;
}

/**
 * Build a {@link Bank} from the authoritative descriptor table in Hod2.exe.
 *
 * Always prefer this over {@link solveLayout}: the exe carries exact offsets,
 * dimensions and layout codes, whereas a prefix sum cannot reproduce the
 * 2048-byte padding or the VQ half-size convention.
 */
export function bankFromExe(data: Uint8Array,
                            entries: readonly TexEntryLike[]): Bank {
  const descs = new Map<number, TexDesc>();
  const offsets = new Map<number, number>();
  for (const e of entries) {
    descs.set(e.index, {
      width: e.width, height: e.height, pixfmt: e.pixfmt,
      vq: e.vq, mipmap: false, twiddled: e.twiddled,
    });
    offsets.set(e.index, e.offset);
  }

  let used = 0;
  if (entries.length) {
    // `max(entries, key=...)` keeps the first of a tie; so does this.
    let last = entries[0];
    for (const e of entries) if (e.offset > last.offset) last = e;
    used = last.offset + descSize(descs.get(last.index)!);
  }
  return { data, descs, offsets, complete: used <= data.length,
           residual: data.length - used };
}

/**
 * Compute the byte offset of each texture in a bank.
 *
 * Textures are concatenated in ID order, so offsets are a prefix sum -- but
 * only if every ID from 0..max is known. A gap (an ID present in the bank but
 * referenced by no model we parsed) makes every later offset unknown.
 *
 * Returns a bank with whatever could be resolved. `complete` is true when the
 * prefix sum covers every ID and lands exactly on the file size.
 */
export function solveLayout(data: Uint8Array,
                            descs: Map<number, TexDesc>): Bank {
  if (descs.size === 0) {
    return { data, descs, offsets: new Map(), complete: false,
             residual: data.length };
  }

  const top = Math.max(...descs.keys());
  const offsets = new Map<number, number>();
  let pos = 0;
  let ok = true;
  for (let i = 0; i <= top; i++) {
    const d = descs.get(i);
    if (d === undefined) { ok = false; break; }
    offsets.set(i, pos);
    pos += descSize(d);
  }

  const residual = data.length - pos;
  return { data, descs, offsets, complete: ok && residual === 0, residual };
}
